transports/tcp/src/lib.rs

```diff
@@ -425,14 +481,111 @@ where
     /// `None` is returned if one of the given addresses is not a TCP/IP
     /// address.
     fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
-        match &self.port_reuse {
+        match &self.config.port_reuse {
             PortReuse::Disabled => address_translation(listen, observed),
             PortReuse::Enabled { .. } => Some(observed.clone()),
         }
     }
+
+    // TODO: docs
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
+        // Return pending events from closed listeners.
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        // We remove each element from `listeners` one by one and add them back.
+        let mut remaining = self.listeners.len();
+        while let Some(mut listener) = self.listeners.pop_back() {
```

Comment:
Could we make `self.listeners` a `FuturesUnordered`?



While we could not remove listeners, we could instead access them in a mutable fashion via [`iter_mut`](https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html#method.iter_mut) and have them resolve instantly instead.

Just a thought. I think this is out of scope for this pull request.

Reply:
Not sure if I am missing something, but how would this work with the listeners being Streams and not Futures? 
But I just saw that there is [futures::stream::SelectAll](https://docs.rs/futures/latest/futures/stream/select_all/struct.SelectAll.html#); maybe we could use that?

> While we could not remove listeners, we could instead access them in a mutable fashion via [`iter_mut`](https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html#method.iter_mut) and have them resolve instantly instead.

How would we know which future belongs to which ListenerId?

Reply:
Discussed out of band.

> But I just saw that there is [futures::stream::SelectAll](https://docs.rs/futures/latest/futures/stream/select_all/struct.SelectAll.html#); maybe we could use that?

:+1: 



> > While we could not remove listeners, we could instead access them in a mutable fashion via [`iter_mut`](https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html#method.iter_mut) and have them resolve instantly instead.
> 
> How would we know which future belongs to which ListenerId?

We would push a wrapper type into the `SelectAll`, e.g.:

```` rust
struct ListenerWrapper {
  id: ListenerId,
  terminate: bool,
  listener: _,
}
```

`ListenerWrapper` would implement `Stream` and if `self.terminate` is `false` delegate to `self.listener` and otherwise return `Poll::Ready(None)`.

Removing a listener is means iterating the list and setting the flag to `true` on the matching one.